Loop termination

FouadMA opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

fixing conflicts with latest rebase

rmetzger committed:
[FLINK-3660] Measure latency and exposes them via a metric

This commit adds the initial runtime support for measuring latency of records going through the system.

I therefore introduced a new StreamElement, called a LatencyMarker.
Similar to Watermarks, LatencyMarkers are emitted from the sources at an configured interval. The default value for the interval is 2000 ms. The emission of markers can be disabled by setting the interval to 0. LatencyMarkers can not "overtake" regular elements. This ensures that the measured latency approximates the end-to-end latency of regular stream elements.

Regular operators (excluding those participating in iterations) forward latency markers if they are not a sink.
Operators with many outputs randomly select one to forward the maker to. This ensures that every marker exists only once in the system, and that repartition steps are not causing an explosion in the number of transferred markers.
If an operator is a sink, it will maintain the last 512 latencies from each known source instance.
The min/max/mean/p50/p95/p99 of each known source is reported using a special LatencyGauge from the sink (every operator can be a sink, if it doesn't have any outputs).

This commit does not visualize the latency in the web interface.
Also, there is currently no mechanism to ensure that the system clocks are in-sync, so the latency measurements will be inaccurate if the hardware clocks are not correct.

This closes #2386

Commit:
[FLINK-2390] Loop Termination Fix

kirill-morozov-epam committed:
[FLINK-4506] [DataSet] Fix documentation of CsvOutputFormat about incorrect default of allowNullValues

- Add test case for CsvOutputFormat

This closes #2477
This closes #2631

Commit:
[FLINK-4564] [metrics] Configurable delimiter per reporter

This closes #2517.

zentol committed:
[FLINK-4720] Implement archived ExecutionGraph

This closes #2577.

vijikarthi committed:
[FLINK-3932] Added ZK ACL configuration for secure cluster setup

This closes #2589.

vijikarthi committed:
[FLINK-4667] Fix for using correct ZK namespace in Yarn deployment

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:
```diff
@@ -204,4 +204,6 @@ void acknowledgeCheckpoint(
 
 	InputGate[] getAllInputGates();
+
+	void tellJobManger(Object msg);
```
Introduce interface, similarly to CheckpointResponder. Then provide one implementaiton that uses akka, for now. We will revisit this after FLIP-6 is merged

senorcarbone replied:
resolved

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -211,6 +212,10 @@
 	private CheckpointCoordinator checkpointCoordinator;
 
+	/** The coordinator used to determine when the execution of the iterations is done */
+	private LoopTerminationCoordinator loopTerminationCoordinator;
```
Make sure this is Serializable (in case this creates an issue with Zookeeper HA)

senorcarbone replied:
resolved!
This is not an issue since only task state is maintained in HA snapshots

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/LoopTerminationCoordinator.java:
```diff
@@ -0,0 +1,282 @@
+ * A coordinator that listen to  stream sources finish state and starts the
+ * process of termination detection. It ends with sending a {@link StopInputTasks} event message to
+ * all input tasks to stop, which will propagate the stop to all successors
+
```
very nice sequence diagram! We will have to remake this in the flip.
Would be nice to add steps/bullets with explanation of the component (lifecycle, dependencies etc.) so that the logic is self-contained in the javadoc of the coordinator.

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/LoopTerminationCoordinator.java:
```diff
@@ -0,0 +1,282 @@
+			return; // it is not right
+		}
+
+		if(terminated){
```
these conditions should never occur. We might have to double check what happens when rollback recovery occurs **during** a termination period.

In any case, merge conditions with || if possible

senorcarbone replied:
Just as a reminder: The LoopTerminationCoordinator might have to subscribe to state changes in the execution graph. In case the execution graph is a enters a 'RESTARTING' phase it will have to reset its state. That's all that is needed

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/LoopTerminationCoordinator.java:
```diff
@@ -0,0 +1,282 @@
+				,this.jobId,taskStatus.getJobID());
+
+		ExecutionJobVertex ejv = allVertices.get(taskStatus.getTaskVertexID());
+		if(nonInputVertecies.containsValue(ejv) || headVertices.contains(ejv)){
```
more clear to read if there is a union set of tasksToAcknowledge

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/LoopTerminationCoordinator.java:
```diff
@@ -0,0 +1,282 @@
+	private void onRoundClosure() {
+		boolean allTasksDone = getAllStatus();
+		if(allTasksDone){
+			if(previousStepDone) {
```
redundant. To remove

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/LoopTerminationCoordinator.java:
```diff
@@ -0,0 +1,282 @@
+		}
+	}
+
+}
```
Really nice implementation, it aligns with the CheckpointCoordinator logic.  As a last comment for the coordinator I would suggest to see how it works out when HA is enabled (Zookeeper). Perhaps you might also need to make transient things that are not needed upon a savepoint.

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/WorkingStatusUpdate.java:
```diff
@@ -0,0 +1,81 @@
+ * An event that is triggered by streaming sources to tell all operators to report their state to the
+ * {@link LoopTerminationCoordinator}
+ */
+public class WorkingStatusUpdate extends AbstractEvent {
```
+1
good that this aligns with the blocking queue type for the FT for loops - remake

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/iterative/termination/WorkingStatusUpdate.java:
```diff
@@ -0,0 +1,81 @@
+
+	public WorkingStatusUpdate(){}
+
+	public WorkingStatusUpdate(long seqNum, String sourceName){
```
remove sourceName, after debugging

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/tasks/StatefulTask.java:
```diff
@@ -62,4 +64,7 @@
 	void notifyCheckpointComplete(long checkpointId) throws Exception;
+
+
+	boolean onLoopTerminationCoordinatorMessage(AbstractLoopTerminationMessage msg) throws IOException, InterruptedException;
```
This is fine now. We just have to sync with Stephan etc. whether the logic there will change when selective recovery is merged

senorcarbone replied:
Add Javadoc

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java:
```diff
@@ -238,6 +243,12 @@ public InputGate getInputGate(int index) {
 
+
+	@Override
+	public void tellJobManger(Object msg){
```
implementation specific to Akka. It is fine though as a first implementation as long as we extract interfaces first

senorcarbone commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/IterativeStream.java:
```diff
@@ -71,14 +71,18 @@ protected IterativeStream(DataStream<T> dataStream, long maxWaitTime) {
+				return feedbackStream;
+			}
+			if(t instanceof  FeedbackTransformation && ((FeedbackTransformation) t).getFeedbackEdges().size()==0){
+				throw new UnsupportedOperationException("An inner iteration should be closed first.");
```
This is temporary, I approve it of course. Just a note here that we will eventually introduce proper scopes in a later FLIP

senorcarbone commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java:
```diff
@@ -83,13 +85,16 @@
 								IOManager ioManager,
 								boolean enableWatermarkMultiplexing) throws IOException {
-
+		this.taskLoopFinalizationListeer = taskLoopFinalizationListeer;
```
spelling: ..Listeer -> ..Listener

senorcarbone commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/LoopTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,214 @@
+ * <pre>
+ *
+ *         Number of Channels = C
+ *
```
woohoo, more ascii art. I like the diagram, but once again just add some short description as well (waiting for completion etc)

senorcarbone commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/LoopTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,214 @@
+		isIdle = false;
+	}
+
+	private void onAllRequestsRecieved() {
```
typo: Recieved -> Received

senorcarbone commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/LoopTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,214 @@
+		}
+	}
+
+	class MapWithValueCounter<K,V> extends HashMap<K,V>{
```
put in utils so that others can reuse it potentially? (after syncing with other commiters)

senorcarbone commented on flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/LoopTerminationITCase.java:
```diff
@@ -0,0 +1,281 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
```
add if possible some javadoc to describe the scenario of the ITCase

senorcarbone replied:
resolved! Nice 👍

senorcarbone reviewed (changes requested):
Great work overall! When all small details are addressed I believe this is ready for a FT rebase and a final review.

senorcarbone commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -128,7 +130,7 @@
 
 	/** The chain of operators executed by this task */
-	private OperatorChain<OUT> operatorChain;
+	OperatorChain<OUT> operatorChain;
```
this is deliberately private. I dont see it being used in package access so it can be reverted

senorcarbone reviewed (approved):
Review Phase 1 is done. Nicely Done!
There are only minor style fixes needed to be done by I approve this after you apply them. 

Here are some pending things to sync with the rest of the community:
- The Inner Loop Restriction should be well understood and approved by everyone. There is no notion of scopes there yet in the Flink API but this restriction forces proper scoping for now, so it is a start.
- We need to double check if selective recovery changes affect any part of the logic here. I do not see any problem but it is good to make sure it is the case.
- Maybe it is better to move utils such as MapWithValueCounter to flink-commons but it is still fine for me there.

senorcarbone commented:
just rebased to current master.
@FouadMA please add missing javadocs

senorcarbone commented:
@FouadMA judging from Travis there is still one issue pending. It seems that job cancels are not completed, which means that they are not handled accordingly by the terminator. You can take a look when u find the time. Good job so far

senorcarbone commented:
thanks for the quick fix @FouadMA . Let's keep an eye on the travis builds for now

senorcarbone commented:
There are another two pending changes 
1. Remove the timeout from the API. It is funny that we missed that part. (If we want to merge this to flink < v.2 we keep a version with the timeout so we dont break the API, that will have basically no effect).
2. Mock the JobTerminationHandler in test environments to instantly kill regular sources that have invoked the termination callback. That will align with the existing test logic

tzulitai committed:
[FLINK-4723] [kafka] Unify committed offsets to Kafka to be next record to process

This closes #2580

tzulitai committed:
[FLINK-4727] [kafka] Set missing initial offset states with starting KafkaConsumer position

With this change, on a clean startup of FlinkKafkaConsumer09, the auto retrieved offsets (either earliest, latest, or an actual
committed offset) from Kafka will also be checkpointed and committed, even if no records are read after the startup.

This closes #2585

senorcarbone commented:
@FouadMA I found a condition in the stream graph that might be affected so when you can take a look: There is `vertexIDtoLoopTimeout` in the `StreamGraph` that is used to determine whether a graph contains loops. If we remove timeouts this needs to be addressed too (if that is not the case already)

uce committed:
[FLINK-4512] [FLIP-10] Add option to persist periodic checkpoints

[FLINK-4509] [FLIP-10] Specify savepoint directory per savepoint
[FLINK-4507] [FLIP-10] Deprecate savepoint backend config

This closes #2608.

uce committed:
[FLINK-4717] Add CancelJobWithSavepoint

- Adds CancelJobWithSavepoint message, which triggers a savepoint
  before cancelling the respective job.
- Adds -s [targetDirectory] option to CLI cancel command:
    * bin/flink cancel <jobID> (regular cancelling)
    * bin/flink cancel -s <jobID> (cancel with savepoint to default dir)
    * bin/flink cancel -s <targetDir> <jobID> (cancel with savepoint to targetDir)

This closes #2609.

zentol committed:
[FLINK-4774] [metrics] Fix scope concatenation in QueryScopeInfo

This closes #2613.

qingdao81 committed:
[FLINK-4771] [avro] Add support for compression to AvroOutputFormat.

This closes #2612

rmetzger committed:
[FLINK-3706] Fix YARN test instability

The most important change in this commit is that the `YarnTestBase.Runner` doesn't do "try {} catch (Throwable t) { fail(t); }" anymore, which doesn't lead to a test failure, because its called outside the main thread.
With the change, all throwables are reported back to the main thread and fail the test there properly (many YARN tests benefit from this change).

This closes #2622

twalthr committed:
[FLINK-4108] [scala] Respect ResultTypeQueryable for InputFormats.

This closes #2619

mxm committed:
[FLINK-4799] re-add 'build-target' symlink on Unix systems

This adds back the link 'build-target' which points to the final build
assembly in 'flink-dist/target/'.

This closes #2620

rmetzger committed:
[hotfix][kafka] Undo DataGenerators changes (use inline kafka producer again

rmetzger committed:
[hotfix][kafka] Backport Kafka09FetcherTest for Kafka010

This closes #2627

twalthr committed:
[FLINK-4791] [table] Fix issues caused by expression reduction

StephanEwen committed:
[FLINK-4373] [cluster management] Introduce AllocationID, ResourceProfile, and AllocatedSlot

These classes are introduced as part of the cluster management rework.

This closes #2630

hzyuemeng1 committed:
[FLINK-4827] [docs] Fix scala Streaming Table example

This closes #2632.

zentol committed:
[hotfix] Fix ArchivedExecutionGraphTest

mxm committed:
[FLINK-4829] protect user accumulators against concurrent updates

rmetzger committed:
[FLINK-3660][hotfix] Add missing default value in ConfigConstants

mxm committed:
[FLINK-4667] add import for config entry

skrishnappa committed:
[FLINK-3931] Implement Transport Encryption (SSL/TLS)

Enabled SSL support for the different network transport within flink.
This feature is documented as task T3 in the "Secure Data Access in Flink" design doc - https://docs.google.com/document/d/1-GQB6uVOyoaXGwtqwqLV8BHDxWiMO2WnVzBoJ8oPaAs

This commit addresses subtasks - FLINK-4324, FLINK-4325, FLINK-4404 and FLINK-4405
For details about how to use this feature, refer to security-ssl.md and the new config values added to config.md

This closes #2518

mxm committed:
[FLINK-4829] snapshot accumulators on a best-effort basis

Heartbeats should not fail when accumulators could not be snapshotted. Instead,
we should simply skip the reporting of the failed accumulator. Eventually, the
accumulator will be reported; at the latest, when the job finishes.

This closes #2649
